src/btree/row_srch.c

```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
     match = skiphigh = skiplow = 0;
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
+         * The algorithm requires that the skip list insert pointer is not re-read during the loop.
+         * Whilst the compiler can change the code in a way that it reads the ins value from memory
+         * again in the following code thus breaking this assumption.
+         *
+         * In addition, a CPU with weak memory ordering, such as ARM and PPC, may speculatively read
+         * an old value. It is not OK and the reason is explained in the following comment.
+         *
+         * Place a read barrier here using atomic read to avoid these issues.
```

Comment:
```WT_ORDERED_READ``` uses fence instructions to issue a read barrier; it is not an atomic read. This line needs to be fixed.